src/triggers/newEVMEvent.ts

```diff
@@ -0,0 +1,141 @@
+  }
+
+  return { rangeStart, rangeEnd, mostStaleContracts };
+};
+
+type ContractsEventsCursor = {
+  [contractAddress: string]: string;
+};
+
+export type EVMEventCursorArgs = {
+  id: string;
+  address: string;
+  startingBlock: string;
+};
+
+export const newEVMEvents: (
+  chainId: ChainId,
+  args: EVMEventCursorArgs[],
+  contractFilters: ContractFilter[],
+  gap?: string,
+  staleFilter?: (contractFilters: ContractFilter[], staleArgIds: string[]) => ContractFilter[],
+  updateCursor?: (rangeEnd: string, staleContractFilters: ContractFilter[]) => { [x: string]: string },
+  limit?: number
+) => Trigger<Cursor> = (
+  chainId: ChainId,
+  args: EVMEventCursorArgs[],
+  contractFilters: ContractFilter[],
+  gap: string = process.env.ETHEREUM_BLOCK_QUERY_GAP!,
+  staleFilter,
+  updateCursor,
+  limit = 5000
+) => {
+  const triggerFunc = async (
+    clients: Clients,
+    cursorJSON = '{}',
+  ): Promise<TriggerOutput> => {
+    const client = clients.evmChain[chainId];
+
+    const cursor: ContractsEventsCursor = JSON.parse(cursorJSON) || {};
+    if (args.length === 0) {
+      return [];
+    }
+    args.forEach(arg => {
+      if (!arg.startingBlock) {
+        throw new Error(`Contract ${arg.id} has no starting block`);
+      }
+
+      if (!cursor[arg.address] && arg.startingBlock) {
+        cursor[arg.address] = arg.startingBlock;
+      }
+    });
+    // not all vars are const
+    // eslint-disable-next-line prefer-const
+    let { rangeStart, rangeEnd, mostStaleContracts } = calculateRange(
+      cursor,
+      gap,
+      limit
+    );
+    // Check for confirmations
+    const latestBlock = BigInt(await client.getLatestBlockNumber());
+    if (rangeEnd > latestBlock - NUMBER_OF_CONFIRMATIONS) {
+      rangeEnd = latestBlock - NUMBER_OF_CONFIRMATIONS;
+    }
+
+    if (rangeStart > rangeEnd) {
+      return [];
+    }
+
+    let staleContractFilters;
+    if (staleFilter){
+      staleContractFilters = staleFilter(contractFilters, mostStaleContracts)
+    } else {
+      staleContractFilters = contractFilters.filter(filter => {
+        return mostStaleContracts.includes(filter.address);
+      });
+    }
+    let newCursorObject = { ...cursor };
+    if (updateCursor){
+      const updatedCursor = updateCursor(rangeEnd.toString(), staleContractFilters);
+      newCursorObject = {
+        ...newCursorObject,
+        ...updatedCursor
+      }
+    } else {
+      staleContractFilters.forEach(filter => {
+        newCursorObject[filter.address] = rangeEnd.toString();
+      });
+    }
+
+    if (staleContractFilters.length === 0) {
+      return [];
+    }
+
+    const newEvents = await client.getEventsFrom(
+      rangeStart.toString(),
+      rangeEnd.toString(),
+      staleContractFilters,
+    );
+    const newCursor = JSON.stringify(newCursorObject);
+    // if (newEvents.length === 0 && mostStaleContracts.length !== args.length) {
+    if (newEvents.length === 0 && mostStaleContracts.length !== limit) {
```

Comment:
Need to test this recursion condition further

Reply:
Ping me if you'd like to pair on setting up some codebase tests :)